cmake/FindHiopCudaLibraries.cmake

```diff
@@ -9,32 +9,35 @@ add_library(hiop_cuda INTERFACE)
 
 find_package(CUDAToolkit REQUIRED)
 
-target_link_libraries(hiop_cuda INTERFACE 
-  culibos
-  nvblas
-  cusparse
-  cusolver
-  cudart
-  cublasLt
+# # Uncomment for old workaround
+# target_link_libraries(hiop_cuda INTERFACE 
+#   culibos
+#   nvblas
+#   cusparse
+#   cusolver
+#   cudart
+#   cublasLt
+#   )
+
+if(HIOP_BUILD_SHARED)
+  target_link_libraries(hiop_cuda INTERFACE 
+    nvblas # shared only
+    CUDA::cusparse
+    CUDA::cudart
+    cublas
+    cublasLt
+    CUDA::cusolver
   )
-
-# if(HIOP_BUILD_SHARED)
-#   target_link_libraries(hiop_cuda INTERFACE 
-#     CUDA::cusparse
-#     CUDA::cudart
-#     CUDA::cublasLt
-#     CUDA::cusolver
-#     )
-# endif()
-# if(HIOP_BUILD_STATIC)
-#   target_link_libraries(hiop_cuda INTERFACE
-#     CUDA::culibos
-#     CUDA::cusparse_static
-#     CUDA::cudart_static
-#     CUDA::cublasLt_static
-#     CUDA::cusolver_static
-#     )
-# endif()
+endif()
+if(HIOP_BUILD_STATIC)
+  target_link_libraries(hiop_cuda INTERFACE
+    CUDA::cusparse_static
+    CUDA::cudart_static
+    cublas
+    cublasLt
+    CUDA::cusolver_static
```

Comment:
@CameronRutherford, is there a reason why some of these library names are prefixed with the namespace `CUDA::` and others are not? It is a bit confusing.

Reply:
See my comment for more context. The workaround that we are using here is to link against either the static *or* shared version of these libraries though the `LD_LIBRARY_PATH`. This seems to resolve the issue of some libraries not finding the right shared/static library call at build time, but I am still not satisfied with it being a good final solution.

Reply:
So cublas and cublasLt can be either shared or static, but we mandate cudart and cusolver be static? Seems like we should allow shared and static for all libraries in that case

Reply:
If it's the case that we want to link against both static and shared variants of these CUDA libraries, then we could incorporate that logic into the CMake directly, rather than inferring it and hoping the linker does the right thing. I imagine declaring `HIOP_LINK_CUDA_SHARED` and `HIOP_LINK_CUDA_STATIC` as advanced CMake options could work.

Part of the issue in this case is the statically built STRUMPACK links against the shared cuda libs in the export CMake, and so the linker has to interpret which one to use at link time (?) to work around this issue. I imagine linking both `CUDA::cublas` and `CUDA::cublas_static` would have the equivalent affect as this fix. Note that this fix seemingly would only apply when building with STRUMPACK.

If we want to unconditionally try and link against both then we can, but perhaps we can decide on a better practice here...